### PR TITLE
Dockerfile template: clarify efa_installer flags, bump versions

### PR DESCRIPTION
**Pull request by verdimrc**
*Issue #, if available:* N/A

*Description of changes:*
- make template Dockerfile to explain the efa_installer flags.
- bump versions

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by verdimrc** [comment]
Note that the base pytorch container `23.12` is built for nvidia driver 545.23 or later.

The container will complain on older driver:

```text
ERROR: This container was built for NVIDIA Driver Release 545.23 or later, but
       version 535.104.12 was detected and compatibility mode is UNAVAILABLE.

       [[]]
```

**Review by perifaws — APPROVED** [review]

File: 2.ami_and_containers/2.docker/0.nvcr-pytorch-aws.dockerfile [hunk 1]
```diff
@@ -63,7 +63,13 @@ RUN apt-get update && \
     curl -O https://efa-installer.amazonaws.com/aws-efa-installer-${EFA_INSTALLER_VERSION}.tar.gz  && \
     tar -xf aws-efa-installer-${EFA_INSTALLER_VERSION}.tar.gz && \
     cd aws-efa-installer && \
-    ./efa_installer.sh -y -g -d --skip-kmod --skip-limit-conf --no-verify && \
+    # Anything from --skip-kmod onwards are used ONLY for container image. Do not use these
```
Comment (mhuguesaws) [inline 2.ami_and_containers/2.docker/0.nvcr-pytorch-aws.dockerfile:66]:
```suggestion
    # ONLY add `--skip-kmod`, `--skip-limit-conf` and `--no-verify` flags to container image.
```

File: 2.ami_and_containers/2.docker/0.nvcr-pytorch-aws.dockerfile [hunk 2]
```diff
@@ -63,7 +63,13 @@ RUN apt-get update && \
...
+    # flags on the host, unless you want to fine-grained control the installation process.
```
Comment (mhuguesaws) [inline 2.ami_and_containers/2.docker/0.nvcr-pytorch-aws.dockerfile:67]:
```suggestion
    # Those three flags must NOT be used on the host.
```

File: 2.ami_and_containers/2.docker/0.nvcr-pytorch-aws.dockerfile [hunk 3]
```diff
@@ -63,7 +63,13 @@ RUN apt-get update && \
...
+    #
+    # NOTE: --skip-kmod and --no-verify in Dockerfile causes docker build to hard fail. The
```
Comment (mhuguesaws) [inline 2.ami_and_containers/2.docker/0.nvcr-pytorch-aws.dockerfile:69]: This sentence is very unclear. Does it mean the command below will lead to container image build failure?

**Review by mhuguesaws — COMMENTED** [review]
Left comments.

File: 2.ami_and_containers/2.docker/0.nvcr-pytorch-aws.dockerfile (hunk 3 above)
Comment (sean-smith) [reply 2.ami_and_containers/2.docker/0.nvcr-pytorch-aws.dockerfile:69]: I think this should read:
> NOTE: to build EFA in the Dockerfile, we added `--skip-kmod` and `--no-verify`. Without these flags, the Dockerfile will fail to build. If installing EFA on the *host* and not in a container, please remove these flags.

**Comment by verdimrc** [comment]
Forward compatibility enabled on p5, so not an issue. [Previous comment](https://github.com/aws-samples/awsome-distributed-training/pull/89#issuecomment-1876261293) was when testing with a small instance of g4 (or g5).

**Pull request merged**